### PR TITLE
test on current python/django versions, linter config

### PR DESCRIPTION
**Pull request by chris48s**
This PR continues the wider platform and version upgrades work I am looking at across DC repos.

Key topics on this one are bumping to current/applicable python/django versions and bringing linter setup into line with other DC projects.

Commit log should be pretty descriptive for this one. I will also leave some comments inline on the diff with a bit of further explanation.

### CODE REVIEW

File: .circleci/config.yml
Comment (chris48s) [inline .circleci/config.yml:49]: Running `ruff format .` in CI isn't really helpful because ruff exits with code 0 even if it changes files, so if files need reformatting, `ruff format .` will just change them and not fail the build.
In CI we need to run `ruff format . --check` which will exit with code 1 if filed need reformatting.

File: pyproject.toml
Comment (chris48s) [inline pyproject.toml:59]: This is the standard config we should be using
See https://docs.google.com/document/d/1Sb_m86KG5ca0lOZMgtJZwch32eVfA0syAY9MeIW1Ljk/edit#heading=h.gnuw7hn63sl7
Note that adding the config here is the reason we now have a load of new errors that need fixing (mostly import sorting)

File: .circleci/config.yml
Comment (chris48s) [inline .circleci/config.yml:41]: Same comment as https://github.com/DemocracyClub/EveryElection/pull/2210#discussion_r1718178871 here for the same reasons

File: dc_utils/tests/test_dc_forms.py
Comment (chris48s) [inline dc_utils/tests/test_dc_forms.py:38]: This never did anything because `called_once_with` isn't a function. Same with the other similar line in this file.

This was not a visible error when running the tests under python 3.10, but python 3.12 introduces a check for this. Related reading:

- https://github.com/python/cpython/issues/100690
- https://discuss.python.org/t/include-prefix-called-in-list-of-forbidden-method-prefixes-for-mock-objects-in-unsafe-mode/22249

File: .circleci/config.yml
Comment (chris48s) [inline .circleci/config.yml:9]: Everything we care about/where we consume this package should be running python 3.10 or 3.12 now.
We don't need to keep testing for python 3.8/3.9 compatibility.

File: .circleci/config.yml
Comment (chris48s) [inline .circleci/config.yml:66]: Everything we care about/where we consume this package should now be running Django 4.2.x. We can drop compatibility testing for everything older than that.
I've also added testing on Django 5.1.x so we're flagging any compatibility issues we are likely to encounter when upgrading to Django 5.2.x after the May 2025 elections.

File: pyproject.toml
Comment (chris48s) [inline pyproject.toml:9]: In general, unless we know we're relying on something we know will break in a future release, I'd tend to avoid upper-bound version constraints for python packages.

**Comment by chris48s** [comment]
TODO: once this is approved, cut a release after merging

**Comment by VirginiaDooley** [comment]
I tested this by installing it locally in Website. Is the plan to remove `pyproject.toml` from upstream repos or is there further configuration needed to designate which `pyproject.toml` should be used in the event that more than one is found?

**Review by VirginiaDooley — APPROVED** [review]

**Comment by VirginiaDooley** [comment]
> I tested this by installing it locally in Website. Is the plan to remove `pyproject.toml` from upstream repos or is there further configuration needed to designate which `pyproject.toml` should be used in the event that more than one is found?

Chris and I had a chat about this and he explained more about how `pyproject.toml`  works so all sorted now.

**Pull request merged**
